src/unauthenticated_stream/chacha20.rs

```diff
@@ -0,0 +1,65 @@
+// Copyright 2018 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
```

Comment:
I think `unauthenticated_encryption` is a better module name than `unauthenticated_stream`.

It seems like a lot (most?) of this code is duplicated from the existing ChaCha20 AEAD code. I would really like to avoid this duplication by reorganizing the code so that the ChaCha20-Poly1305 AEAD code is implemented in terms of this new unauthenticated API as much as possible. Basically, I would expect this new API to replace the private API in chacha20.rs, i.e. starting with `git mv src/chacha20.rs src/unauthenticated_encryption/`. See below for specific examples.

If this is hard to do, let's discuss the difficulties and see if we can resolve them together without duplicating the code.

Reply:
One problem of reusing `unauthenticated_encryption::chacha20` as part of `aead::chacha20_poly1305` is that, if I understand things correctly, it would require making `unauthenticated_encryption::chacha20` public. I think it would be better to keep `chacha20` and move the additional common bits and pieces to that instead, so we can keep them private, thoughts?

Reply:
> Also, would need another private module to make common functions and macros shared by `unauthenticated_encryption` and `aead` (that is, unrelated to chacha itself) internally public. Though maybe I'm missing something about how Rust public/private modules work.

I think `pub(crate)` would probably work. It's hard for me to say without trying it.

The important thing is to minimize duplicated logic.

Reply:
Done.